### PR TITLE
Refactor AzureTreeDataProvider so that IAzureNode and IAzureTreeItem are one object

### PR DESCRIPTION
**Pull request by EricJizbaMSFT**
You should never have to do `node.treeItem` again! I completely got rid of 'node' as a term in this codebase. I have coalesced on 'tree item'.

I also moved all of the subscription information into a 'root' object on the TreeItem. This was so that custom root nodes (like the "Attached Accounts" node in Cosmos DB) could have their own root data that they pass down (since they don't have subscription data).

I used storage as a test subject since it has one of the simpler trees. PR for that coming soon.

### CODE REVIEW

File: ui/src/treeDataProvider/AzureNode.ts [hunk 1]
```diff
@@ -3,143 +3,117 @@
  *  Licensed under the MIT License. See License.txt in the project root for license information.
```
Comment (nturinski) [inline ui/src/treeDataProvider/AzureNode.ts]: FIle name is AzureNode.ts despite the class being AzureTreeItem

File: ui/index.d.ts [hunk 2]
```diff
@@ -17,157 +17,257 @@ export type OpenInPortalOptions = {
...
+ * Implement this class if your tree item does not have children. Otherwise use AzureParentTreeItem
+ * NOTE: *Impl methods are not meant to be called directly - just implemented.
  */
-export interface IAzureTreeItem {
+export declare abstract class AzureTreeItem<T = ISubscriptionRoot> {
+    //#region Properties implemented by base class
     /**
      * This is is used for the openInPortal action. It is also used per the following documentation copied from VS Code:
-     * Optional id for the tree item that has to be unique across tree. The id is used to preserve the selection and expansion state of the tree item.
+     * Optional id for the tree treeItem that has to be unique across tree. The id is used to preserve the selection and expansion state of the tree treeItem.
```
Comment (nturinski) [inline ui/index.d.ts]: 🌲 tree treeItem 🌳

File: ui/src/treeDataProvider/AzureNode.ts (hunk 1 above)
Comment (nturinski) [reply ui/src/treeDataProvider/AzureNode.ts]: Never mind, I see your todo note now

File: ui/src/treeDataProvider/AzureParentNode.ts [hunk 3]
```diff
@@ -137,44 +146,42 @@ export class AzureParentNode<T extends IAzureParentTreeItem = IAzureParentTreeIt
             this._cachedChildren = [];
         }
 
-        const sortCallback: (n1: IAzureNode, n2: IAzureNode) => number =
-            this.treeItem.compareChildren
-                ? this.treeItem.compareChildren
-                : (n1: AzureNode, n2: AzureNode): number => n1.treeItem.label.localeCompare(n2.treeItem.label);
+        const sortCallback: (n1: AzureTreeItem<T>, n2: AzureTreeItem<T>) => number =
```
Comment (nturinski) [inline ui/src/treeDataProvider/AzureParentNode.ts]: n1 and n2, huh? Not ti? 😛

Comment (EricJizbaMSFT) [reply ui/src/treeDataProvider/AzureParentNode.ts]: If you wanted to convince me you're actually reviewing this whole PR - you've succeeded haha

File: ui/src/treeDataProvider/InternalInterfaces.ts [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import { AzureTreeItem } from './AzureNode';
+
+// Interfaces for methods on the tree that aren't exposed outside of this package
+// We can't reference the classes directly because it would result in circular dependencies
+
+export interface IAzureParentTreeItemInternal<T = ISubscriptionRoot> extends AzureParentTreeItem<T>, AzureTreeItem<T> {
+    parent: IAzureParentTreeItemInternal<T> | undefined;
+    treeDataProvider: IAzureTreeDataProviderInternal<T>;
+    removeChildFromCache(node: AzureTreeItem<T>): Promise<void>;
+    loadMoreChildren(): Promise<void>;
```
Comment (nturinski) [inline ui/src/treeDataProvider/InternalInterfaces.ts]: Why is this here?

File: ui/README.md [hunk 5]
```diff
@@ -46,13 +46,14 @@ actionHandler.registerEvent('yourExtension.onDidSaveTextDocument', vscode.worksp
...
-    export class WebAppTreeItem implements IAzureTreeItem {
+    export class WebAppTreeItem extends AzureTreeItem {
         public static contextValue: string = 'azureWebApp';
         public readonly contextValue: string = WebAppTreeItem.contextValue;
         private readonly _site: Site;
-        constructor(site: Site) {
+        constructor(parent: AzureParentTreeItem, site: Site) {
+            super(parent);
             this._site = site;
         }
```
Comment (nturinski) [inline ui/README.md]: Is `get id()` and `get label()` still accurate for the readme?

Comment (EricJizbaMSFT) [reply ui/README.md]: Yes

File: ui/src/treeDataProvider/RootNode.ts [hunk 6]
```diff
@@ -3,19 +3,29 @@
...
+        super(undefined);
+        this._root = root;
     }
 
-    public get treeDataProvider(): AzureTreeDataProvider {
+    public get treeDataProvider(): IAzureTreeDataProviderInternal<T> {
         return this._treeDataProvider;
     }
+
+    public set treeDataProvider(treeDataProvider: IAzureTreeDataProviderInternal<T>) {
```
Comment (nturinski) [inline ui/src/treeDataProvider/RootNode.ts]: Why do we need to set this after construction? (as we used to just set it in the constructor)
